### PR TITLE
feat: expose proposal api to calimero sdk

### PR DESCRIPTION
**Pull request by MatejVukosav**
Exposed wasm `create_proposal` function which creates proposal in proxy contract

### CODE REVIEW

File: crates/server/src/admin/service.rs
Comment (xilosada) [inline crates/server/src/admin/service.rs:154]: why not just return the size in proposals? If you have all you just use `len` and if they come paginated it should come in the response

**Review by xilosada — COMMENTED** [review]
:shipit:

File: crates/server/src/admin/handlers/proposals.rs
Comment (xilosada) [inline crates/server/src/admin/handlers/proposals.rs]: why not `Json` instead?

File: crates/server/src/admin/service.rs
Comment (MatejVukosav) [reply crates/server/src/admin/service.rs:154]: pagination will be second iteration

File: crates/server/src/admin/service.rs
Comment (xilosada) [inline crates/server/src/admin/service.rs:25]: TBH I don't understand how this lint rules works 😓

**Comment by MatejVukosav** [comment]
test

File: crates/server/src/admin/handlers/proposals.rs
Comment (petarjuki7) [inline crates/server/src/admin/handlers/proposals.rs:124]: I think `to_string()` should be everywhere where you are converting a `&str` to `String`, since it's more clear what you are trying to do, but might be just a nitpick, not sure about performance (cc: @miraclx)

File: crates/server/src/admin/handlers/proposals.rs
Comment (petarjuki7) [inline crates/server/src/admin/handlers/proposals.rs:131]: same here

File: crates/server/src/admin/handlers/proposals.rs
Comment (petarjuki7) [inline crates/server/src/admin/handlers/proposals.rs:151]: same here... and the others

File: crates/server/src/admin/handlers/proposals.rs
Comment (petarjuki7) [inline crates/server/src/admin/handlers/proposals.rs:14]: Should be split into separate files, and most of the structs into the `primitives` crate

**Review by petarjuki7 — CHANGES_REQUESTED** [review]

File: crates/server/src/admin/handlers/proposals.rs
Comment (MatejVukosav) [reply crates/server/src/admin/handlers/proposals.rs:124]: It was before clippy complaint

File: crates/server/src/admin/handlers/proposals.rs
Comment (MatejVukosav) [reply crates/server/src/admin/handlers/proposals.rs:14]: Thats what todo says 😄

**Comment by petarjuki7** [comment]
I added CLI capability, you can test it with:
`cargo run -p meroctl -- --node-name <node_name> --home data --output-format json proxy get num-proposal-approvals <context_id> <proposal_id>`

Other options are:
`proposal`
`proposals`
`num-active-proposals`
`proposal-approvers`

Also had to make some changes in the handlers for the get APIs.

File: crates/sdk/src/env/ext.rs
Comment (MatejVukosav) [inline crates/sdk/src/env/ext.rs:151]: AsRef

File: apps/blockchain/src/lib.rs
Comment (MatejVukosav) [inline apps/blockchain/src/lib.rs:119]: missed asref

File: packages/calimero-sdk/src/rpc/jsonrpc.ts
Comment (MatejVukosav) [inline packages/calimero-sdk/src/rpc/jsonrpc.ts]: jsonrpc

**Review by chefsale — APPROVED** [review]

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs:94]: Why not?
```suggestion
        request: GetProposalMessagesRequest, //I cannot to this??
```

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs]:
```suggestion
        let proposal_id = env::ext::ProposalId(Self::string_to_u8_32(request.proposal_id.as_str()));
```

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs:95]:
```suggestion
```

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs:112]:
```suggestion
        request: SendProposalMessageRequest, //I cannot to this?? How to use camelCase?
```

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs:116]:
```suggestion
        let proposal_id = env::ext::ProposalId(Self::string_to_u8_32(request.proposal_id.as_str()));
```

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs:122]:
```suggestion
                messages.push(request.message);
```

File: apps/blockchain/src/lib.rs
Comment (petarjuki7) [inline apps/blockchain/src/lib.rs:126]:
```suggestion
                let messages = vec![request.message];
```

**Review by petarjuki7 — CHANGES_REQUESTED** [review]

File: apps/blockchain/src/lib.rs
Comment (MatejVukosav) [reply apps/blockchain/src/lib.rs:112]: this does not work

File: apps/blockchain/Cargo.toml
Comment (miraclx) [inline apps/blockchain/Cargo.toml]: `serde` is reexported through `calimero_sdk` as `calimero_sdk::serde`

File: apps/blockchain/Cargo.toml
Comment (miraclx) [inline apps/blockchain/Cargo.toml:21]: rust should warn on this, profiles should only be in the root manifest

File: Cargo.toml
Comment (miraclx) [inline Cargo.toml:35]: can we try a better name maybe?

File: apps/blockchain/src/lib.rs
Comment (miraclx) [inline apps/blockchain/src/lib.rs]:
```suggestion
use calimero_sdk::serde::{Deserialize, Serialize};
```

File: apps/blockchain/src/lib.rs
Comment (miraclx) [inline apps/blockchain/src/lib.rs:29]: this?

```suggestion
    ProposalCreated { id: ProposalId },
```

File: apps/blockchain/src/lib.rs
Comment (miraclx) [inline apps/blockchain/src/lib.rs]: this doesn't work, if you want to log, use `env::log`

File: apps/blockchain/src/lib.rs
Comment (miraclx) [inline apps/blockchain/src/lib.rs]: wanna try returning `ProposalId` instead?

File: apps/blockchain/src/lib.rs
Comment (miraclx) [inline apps/blockchain/src/lib.rs:95]: why is it a struct?

File: apps/blockchain/src/lib.rs
Comment (miraclx) [inline apps/blockchain/src/lib.rs:134]: there's no guarantee a proposal Id will be safely encodable as utf8, use bs58 instead

File: crates/context/config/src/client/env/proxy/mutate/methods.rs
Comment (miraclx) [inline crates/context/config/src/client/env/proxy/mutate/methods.rs:27]: 👍🏽

File: crates/context/src/lib.rs
Comment (miraclx) [inline crates/context/src/lib.rs]:
```suggestion
        let signer_id = signer_id.rt()?;
```

you can read `signing_key` from the `ContextIdentity` record in store

for example, see how this is done when inviting members to a context

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs]: ?

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs]:
```suggestion
        for (proposal_id, actions) in &outcome.proposals {
            let _ = self
                .ctx_manager
                .propose(
                    context_id,
                    executor_public_key,
                    proposal_id,
                    actions,
```

File: crates/context/src/lib.rs
Comment (miraclx) [inline crates/context/src/lib.rs]: since it's going to be serialized, take a `&[u8]` here

File: crates/sdk/src/env.rs
Comment (miraclx) [inline crates/sdk/src/env.rs]: no need for these

File: crates/sdk/src/sys.rs
Comment (miraclx) [inline crates/sdk/src/sys.rs]: you shouldn't need these either

File: crates/sdk/src/env/ext.rs
Comment (miraclx) [inline crates/sdk/src/env/ext.rs]:
```suggestion
    pub const fn approve(self, proposal_id: ProposalId) {
        unsafe { sys::approve_proposal(BufferMut::new(&proposal_id)) }
```

File: crates/sdk/src/env/ext.rs
Comment (miraclx) [inline crates/sdk/src/env/ext.rs]: `DraftProposal::approve` is implied on `send`, no need for this here

File: crates/server/src/admin/handlers/proposals.rs
Comment (miraclx) [reply crates/server/src/admin/handlers/proposals.rs:124]: `str::to_string` [calls](https://github.com/rust-lang/rust/blob/27e38f8fc7efc57b75e9a763d7a0ee44822cd5f7/library/alloc/src/string.rs#L2795) `String::from(&str)` which [calls](https://github.com/rust-lang/rust/blob/27e38f8fc7efc57b75e9a763d7a0ee44822cd5f7/library/alloc/src/string.rs#L2880) `str::to_owned`, all of which are inlined, so in most cases, it should compile to the same code.

I personally prefer just calling `.to_owned()`

File: Cargo.toml
Comment (MatejVukosav) [reply Cargo.toml:35]: that will be in separate repo. here just for testing

File: crates/context/src/lib.rs
Comment (MatejVukosav) [inline crates/context/src/lib.rs:1024]: q

File: apps/blockchain/src/lib.rs
Comment (MatejVukosav) [reply apps/blockchain/src/lib.rs:95]: To match client request as it is cleaner to change request objects instead of method signature.

File: crates/sdk/src/env.rs
Comment (MatejVukosav) [reply crates/sdk/src/env.rs]: why?

File: apps/blockchain/Cargo.toml
Comment (frdomovic) [reply apps/blockchain/Cargo.toml]: Removed

File: apps/blockchain/src/lib.rs
Comment (frdomovic) [reply apps/blockchain/src/lib.rs]: Updated

File: crates/sdk/src/env.rs
Comment (frdomovic) [reply crates/sdk/src/env.rs]: removed

File: crates/sdk/src/env/ext.rs
Comment (frdomovic) [reply crates/sdk/src/env/ext.rs]: Removed

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs]: this should be from `outcome.proposals` which is the outcome of the execution, not `payload` (which is the argument to the function call)

`outcome.proposals` is a `Map<Id, Vec<u8>>`, the `proposal` payload should be deserialized `Vec<ProposalAction>`, and a `Proposal` can be constructed.

a similar path for `outcome.approvals`

Comment (miraclx) [reply crates/node/src/lib.rs]: ah, this is trying to determine the `sender` and `receiver`

`sender` is `executor_public_key`, and `receiver` is determined by the action construction inside the app.

~all that needs to happen here is proposal construction (wrapping already constructed actions)~

I see that `ctx_manager.propose` takes `Vec<Action>`, so this is further simplified

File: crates/sdk/src/env.rs
Comment (miraclx) [reply crates/sdk/src/env.rs]: the `External` API is how developers will interface with these.

what gets passed here is delicate since it'll attempt to be deserialized on the node, so controlling low-level access is ideal

File: crates/sdk/src/env/ext.rs
Comment (miraclx) [inline crates/sdk/src/env/ext.rs:78]: this is needed here

File: apps/blockchain/src/lib.rs
Comment (MatejVukosav) [reply apps/blockchain/src/lib.rs:29]: app will be extracted so tbd

File: crates/context/config/src/client/protocol/near.rs
Comment (miraclx) [inline crates/context/config/src/client/protocol/near.rs:3]: ?

File: crates/context/config/src/client/protocol/near.rs
Comment (miraclx) [inline crates/context/config/src/client/protocol/near.rs:276]: highlighting this to come back to it

File: crates/context/src/lib.rs
Comment (miraclx) [inline crates/context/src/lib.rs:1060]:
```suggestion
            .propose(proposal_id, signer_id.rt().expect("infallible conversion"), actions)
```

while we want to avoid panic points, any thing that is "safe" to use unwrap, must define why that's the case

File: crates/context/src/lib.rs
Comment (miraclx) [inline crates/context/src/lib.rs:1093]: same here

File: crates/meroctl/src/cli.rs
Comment (miraclx) [inline crates/meroctl/src/cli.rs:54]: bad merge? this reintroduces `JsonRpc`

File: crates/node/Cargo.toml
Comment (miraclx) [inline crates/node/Cargo.toml:1]: nit: order

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs:9]: ?

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs:429]: this shouldn't be `clone`d

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs:458]: handle the result instead of `drop`ping.

File: crates/node/src/lib.rs
Comment (miraclx) [inline crates/node/src/lib.rs:466]: same here

File: crates/node/src/types.rs
Comment (miraclx) [inline crates/node/src/types.rs:62]: now unused?

File: crates/sdk/src/env/ext.rs
Comment (miraclx) [inline crates/sdk/src/env/ext.rs:39]: deposit in function call should be `u128` too

File: crates/sdk/src/env/ext.rs
Comment (miraclx) [reply crates/sdk/src/env/ext.rs:78]: 🤦🏽 not needed

**Pull request merged**
